cpp/include/rapidsmpf/streaming/core/channel.hpp

```diff
@@ -6,76 +6,157 @@
- * @tparam Capacity Maximum number of messages the channel can buffer.
+ * @brief Move-only, type-erased message holding a payload as shared pointer.
  */
-template <typename T, std::size_t Capacity = 1>
-class Channel {
+class Message {
   public:
+    Message() = default;
+
     /**
-     * @brief Asynchronously send a value into the channel.
+     * @brief Construct from a unique pointer (promoted to shared_ptr).
      *
-     * Suspends if the buffer is empty.
+     * @tparam T Payload type.
+     * @param ptr Non-null unique pointer.
+     * @throws std::invalid_argument if @p ptr is null.
+     */
+    template <typename T>
+    Message(std::unique_ptr<T> ptr) {
+        RAPIDSMPF_EXPECTS(ptr != nullptr, "nullptr not allowed", std::invalid_argument);
+        data_ = std::shared_ptr<T>(std::move(ptr));
+    }
+
+    /** @brief Move construct. @param other Source message. */
+    Message(Message&& other) noexcept = default;
+
+    /** @brief Move assign. @param other Source message. @return *this. */
+    Message& operator=(Message&& other) noexcept = default;
+    Message(Message const&) = delete;
+    Message& operator=(Message const&) = delete;
+
+    /**
+     * @brief Reset the message to empty.
+     */
+    void reset() noexcept {
+        return data_.reset();
+    }
+
+    /**
+     * @brief Returns true when no payload is stored.
      *
-     * @param value The value to send.
-     * @return A coroutine that evaluates to true if the value was successfully sent or
-     * false if the channel was shut down.
+     * @return true if empty, false otherwise.
      */
-    coro::task<bool> send(T value) {
-        auto result = co_await rb_.produce(std::move(value));
-        co_return result == coro::ring_buffer_result::produce::produced;
+    [[nodiscard]] bool empty() const noexcept {
+        return !data_.has_value();
     }
 
     /**
-     * @brief Asynchronously receive a value from the channel.
+     * @brief Compare the payload type.
      *
-     * Suspends if the buffer is empty.
+     * @tparam T Expected payload type.
+     * @return true if the payload is `T`, false otherwise.
+     */
+    template <typename T>
+    [[nodiscard]] bool holds() const noexcept {
+        return data_.type() == typeid(std::shared_ptr<T>);
+    }
+
+    /**
+     * @brief Extracts the payload and resets the message.
      *
-     * @return A coroutine that evaluates to an optional containing the received value or
-     * std::nullopt if the channel is shut down.
+     * @tparam T Payload type.
+     * @return The payload.
+     * @throws std::invalid_argument if empty or type mismatch.
      */
-    coro::task<std::optional<T>> receive() {
-        auto msg = co_await rb_.consume();
-        if (msg.has_value()) {
-            co_return std::make_optional(std::move(*msg));
-        } else {
-            co_return std::nullopt;
-        }
+    template <typename T>
+    T release() {
+        auto ret = get_ptr<T>();
+        reset();
```

Comment:
IIUC, if some other message also has a reference to this payload they continue to do so?

Reply:
Thanks.